clippy_lints/src/redundant_test_prefix.rs

```diff
@@ -0,0 +1,143 @@
+use clippy_config::Conf;
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::visitors::for_each_expr;
+use clippy_utils::{is_in_cfg_test, is_test_function};
+use rustc_errors::Applicability;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{self as hir, Body, ExprKind, FnDecl};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::impl_lint_pass;
+use rustc_span::Span;
+use rustc_span::def_id::LocalDefId;
+use std::ops::ControlFlow;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for test functions (functions annotated with `#[test]`) that are prefixed
+    /// with `test_` which is redundant.
+    ///
+    /// ### Why is this bad?
+    /// This is redundant because test functions are already annotated with `#[test]`.
+    /// Moreover, it clutters the output of `cargo test` since test functions are expanded as
+    /// `module::tests::test_use_case` in the output. Without the redundant prefix, the output
+    /// becomes `module::tests::use_case`, which is more readable.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// #[test]
+    /// fn test_use_case() {
+    ///     // test code
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// fn use_case() {
+    ///    // test code
+    /// }
+    /// ```
+    #[clippy::version = "1.84.0"]
+    pub REDUNDANT_TEST_PREFIX,
+    pedantic,
```

Comment:
Copy of the discussion from the previous PR: https://github.com/rust-lang/rust-clippy/pull/12861/files#r1625940772

This should be decided on during the FCP

---

> @xFrednet
> I feel like this shouldn't be in `pedantic` but in `restriction` instead. The usage of the `test_` prefix doesn't seem like an antipattern to me. There are good arguments against it, but the number of changes required in existing changes shows that the usage is still quite common.
> 
> I'll include this decision in the group discussion before we merge this PR :)


> @Caylub
> There are arguments to be made for several categories: `style`, `pedantic`, `restriction`, etc.
> 
> The usage of `test_` seems to be `pedantic` to me, because it's clearly unnecessary in almost every case, including some of the ones mentioned in the threads in this PR. For example, test names that have `test_` + `method_name`, should just be `super::method_name` IMO, following the same like of reasoning for as everywhere else.


> @xFrednet
> One argument I always consider is the number of lint triggers that happen in current code. Clippy can be a very pedantic, but the number of lint emissions has to be reasonable. If users get 100+ warnings of this lint with the next Clippy release it's way more likely that users get annoyed and simply allow the lint and that would help nobody.
> 
> This is not meant as a hard no, but just an explanation of my suggestion. I'm not sure where to place it and would leave this up to a group decision.

Reply:
Have no strong opinion on this one (used `pedantic` just as a starting point). Can wait for the group discussion, or can proactively turn it to `restriction`, since potentially this can indeed result in 100s of warnings on big projects, and that will annoy ppl.